docs/DesignDocs/CustomAttestation.md

```diff
@@ -646,9 +646,29 @@ typedef void* oe_sgx_plugin_opt_params;
  * Helper function that returns the SGX attester that can then be sent to
  * `oe_register_attester`.
  *
+ * @deprecated Instead, use oe_sgx_plugin_local_attester()
```

Comment:
Please update this whole document, or better yet, submit a new design document that highlights the differences between the previous custom attestation model and the proposed changes. The API updates here are insufficient and confusing as they contradict things elsewhere in the doc (e.g. this doc should probably not have any SGX plugin information, that can just be looked up in code directly rather than keeping this in sync with actual implementation)

Things I'm looking to be explained:
* Why is there the shift to splitting local & remote attesters into separate implementations?
* What does it mean for a dev? Do I always register both? Why can I use attestation (e.g. samples) without explicitly registering either today? (i.e. is there a fallback to default sgx plugins even if I don't register them but register some other plugins?)
 
A sample is really necessary here and if there is not one already, please file an issue to add one.

Reply:
I think local vs remote should NOT be split into separate APIs that enclave apps call, as I think in general there is not a clean distinction between those terms.  We don't want SGX-specific things in a generic API.  (It's fine if it's in plugin-internal code.)

Reply:
Thinking more, I think even OE_EVIDENCE_FLAGS_REMOTE_ATTESTATION is problematic, and I haven't seen a definition of "local" or "remote" in this document.   E.g., if you're attesting between an enclave and a USB-attached peripheral, is that local or remote?   I think most users would consider that local.

Reply:
There are 3 ways to get SGX evidences (reports) - local, dcap remote, and EPID remote. Each needs different sets of input for the report to be generated. Application has to be aware of the type of attestation and pass in parameters accordingly.

---
In reply to: [384113533](https://github.com/openenclave/openenclave/pull/2579#discussion_r384113533) [](ancestors = 384113533)

Reply:
SGX uses terms "report" and "quote" while OE uses "local report" and "remote report" in attestation API. To follow the same terminology, I think we should continue to use local. The remote will be replaced with "DCAP". More SGX remote attestation format such as EPID will join the plugin family.

---
In reply to: [384119063](https://github.com/openenclave/openenclave/pull/2579#discussion_r384119063) [](ancestors = 384119063)

Reply:
> There are 3 ways to get SGX evidences (reports) - local, dcap remote, and EPID remote. 
> Each needs different sets of input for the report to be generated. Application has to be aware of the type of attestation and pass in parameters accordingly.

Where is that documented in this document, and how is an application supposed to know which to use?   Where is the sample that works with all three that shows how it knows what to use?  (This seems like a bad model to me.)

Reply:
I see the "flags" in oe_get_evidence() as a carry-over from oe_get_report(), which does not take in evidence format UUID. For SGX, oe_get_report() only generates two types of reports: local, or remote (which is ECDSA), as set by the “flags” parameter.

In oe_get_evidence(), UUID is used to set the format of the evidence to be generated. For SGX, UUID could be OE_SGX_LOCAL_ATTESTATION_UUID, 
OE_SGX_ECDSA_P256_ATTESTATION_UUID, and other formats as their support is added. With the UUID, the “flags” parameter becomes redundant and could even conflict with the UUID parameter. For example, in oe_get_evidence_v2(), what if the UUID is OE_SGX_ECDSA_P256_ATTESTATION_UUID while the “flags” is OE_EVIDENCE_FLAGS_LOCAL_ATTESTATION? It becomes impossible for oe_get_evidence_v2() to honor both parameters. Comments for this PR suggest that “flags” should take precedence.

New version oe_get_evidence() removes the carry-over and resolves the conflict by removing the redundant “flags” parameter.